rust/src/settings.rs

```diff
@@ -0,0 +1,184 @@
+//! Module for loading and validating coordinator settings.
+//!
+//! Settings defined in the configuration file can be overridden by environment variables.
+use crate::mask::{BoundType, DataType, GroupType, ModelType};
+use config::{Config, ConfigError, Environment};
+use thiserror::Error;
+use validator::{Validate, ValidationErrors};
+
+#[derive(Error, Debug)]
+pub enum SettingsError {
+    #[error("configuration loading failed: {0}")]
+    Loading(#[from] ConfigError),
+    #[error("validation failed: {0}")]
+    Validation(#[from] ValidationErrors),
+}
+
+#[derive(Debug, Validate, Deserialize)]
+#[doc(hidden)]
+pub struct Settings {
+    #[validate]
+    pub api: ApiSettings,
+    #[validate]
+    pub pet: PetSettings,
+    pub mask: MaskSettings,
+}
+
+#[derive(Debug, Validate, Deserialize)]
+/// PET protocol settings
+pub struct PetSettings {
+    #[validate(range(min = 1))]
+    pub min_sum: usize,
+    #[validate(range(min = 3))]
+    pub min_update: usize,
+
+    #[validate(range(min = 0, max = 1.0))]
```

Comment:
do you know, for example from testing, if the bounds on `range` are inclusive or exclusive or partly, meaning `min <= value <= max` or `min < value < max` or `min <= value < max`? the documentation doesn't say anything about this. if it were exclusive, we could already rule out corner cases.

especially for the integer values of the `PetSettings` this is quite important.

Reply:
i had a look into the `validator` source code (i should have done this right away^^): the bounds are inclusive, i.e. it is always checked that `min <= value <= max`. also the data type doesn't matter, everything is treated as `f64`.

btw the same holds true for the bounds of the `length` check. this makes sense in the validation case, but it is quite counter intuitive from the usual behavior of `Range<_>`.